### PR TITLE
Add support for post-logging

### LINKED ISSUE #24: Add support for post-logging
In ecAudit we are using pre-logging by logging an ATTEMPT before passing the operation down to the QueryProcessor. Further, in the event of failure we will also log a matching FAILURE message.

In C* 4.0 we took a different approach. One message will always be logged _after_ the operation has completed and at the same time we're indicating if operation was successful or not.

Upside of pre-logging is that we will have a audit event even if a request triggers a fatal stop of the Cassandra node. Downside is the possibly double messages (performance and sometimes unexpected behavior).

We should add support for post-logging to align with Cassandra 4.0, and make that behavior default.

### PR DESCRIPTION
**Pull request by etedpet**
By default ecAudit use pre-logging by logging an ATTEMPT before passing
the operation down to the QueryProcessor. In the event of a failure, a FAILURE
message is also logged.

With post-logging only one message (SUCCEEDED or FAILED) will be logged after
the operation has completed. This is similar to the C* 4.0 approach.

Closes #24

### CODE REVIEW

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/AuditAdapter.java [hunk 1]
```diff
@@ -38,6 +38,7 @@
 import org.apache.cassandra.service.ClientState;
 import org.apache.cassandra.utils.FBUtilities;
 import org.apache.cassandra.utils.MD5Digest;
+import org.jetbrains.annotations.NotNull;
```
Comment (etedpet) [inline ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/AuditAdapter.java]: remove

File: ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/config/TestAuditYamlConfigurationLoader.java [hunk 2]
```diff
@@ -86,6 +86,7 @@ public void testEmptyWhitelistIsStillValid()
         AuditConfig config = givenLoadedConfig(properties);
 
         assertThat(config.getYamlWhitelist()).isEmpty();
+        assertThat(config.isPostLogging()).isFalse();
```
Comment (eperott) [inline ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/config/TestAuditYamlConfigurationLoader.java]: This test should probably sit on the tc above (or both)

File: CHANGES.md [hunk 3]
```diff
@@ -10,6 +10,7 @@
 * Build with Cassandra 3.0.18 (only flavor ecaudit_c3.0)
 * Introduce configurable log message format - #55
 * Make the audit whitelist table a protected resource in Cassandra
+* Add support for post-logging - #24
```
[the remark this answers is not in the document]
Comment (etedpet) [reply CHANGES.md]: I noticed that in ecChronos they where added at the end of the section... So I thought that was where to put it ;-)
But I'll change to the top!

File: doc/setup.md [hunk 4]
```diff
@@ -32,6 +32,10 @@ Two different logger backends are supported out of the box:
 * Then there is the [Chronicle Logger](chronicle_logger.md) backend which has the best performance characteristics.
   This backend stores audit records in a binary format and is best suited when handling large volumes of records.
 
+### Logger Timing
+
+Logger timing specifies *when* log entries should be written, **pre-logging** (legacy) and **post-logging** (C* 4.0 style) are available.
```
[the remark this answers is not in the document]
Comment (etedpet) [reply doc/setup.md]: done

File: doc/description.md [hunk 5]
```diff
@@ -9,6 +9,9 @@ It is possible to create a whitelist of usernames/resources which will not be in
 ecAudit will create an audit record before the operation is performed, marked with status=ATTEMPT.
```
[the remark this answers is not in the document]
Comment (etedpet) [reply doc/description.md]: done

File: conf/audit.yaml [hunk 6]
```diff
@@ -28,6 +28,17 @@
 #    - username2
 
 
+# logger_timing - specify WHEN ecAudit should write log entries. The two options are:
+#
+# pre_logging  -> Will log an ATTEMPT before passing the operation to the QueryProcessor.
+#                 In the event of failure a matching FAILED message will also be logged.
+#                 This is the default value - to be backwards compatible.
```
[the remark this answers is not in the document]
Comment (etedpet) [reply conf/audit.yaml]: done

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/LoggingStrategy.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+{
+    /**
+     * @param status the log operation status
+     * @return {@code true} if the provided status should be logged, {@code false} otherwise.
+     */
+    boolean logStatus(Status status);
+
+    /**
+     * @param status the log operation status
+     * @return {@code true} if the only a summary should be logged for the provided status, {@code false} otherwise.
```
[the remark this answers is not in the document]
Comment (etedpet) [reply ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/LoggingStrategy.java]: removed the Status parameter to this method. It sounded strange to have a method shouldLogFAILEDBatchSummary and then send in a status that might not be failed... Better to have that logic outside the strategy

File: ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/config/TestAuditYamlConfigurationLoader.java (hunk 2 above)
Comment (etedpet) [reply ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/config/TestAuditYamlConfigurationLoader.java]: true, fits better there

**Pull request merged**